### DISCUSSION TITLE
Errors object is empty in tests after upgrading to v6

### DISCUSSION THREAD

**Discussion by MartinCerny-awin** [react-hook-form/react-hook-form#2168] [Q&A] [answered]
I have couple of tests and I have recently upgraded to version 6. The one that got broken is rules validation. Errors is always empty. I have asked the similar question on Spectrum https://spectrum.chat/react-hook-form/help/form-data-and-errors-are-empty-in-react-native~db6a0446-1ec6-40d3-abe6-91d7f04e8c52 and you have suggested me to add

```  
beforeAll(() => {
    global.window = {};
  });
  afterAll(() => {
    global.window = undefined;
  });
```
which worked in version 5, but it does not have any effect in version 6. I think it is because it got recently added to setup file https://github.com/react-hook-form/react-hook-form/blob/1441a0186b8eab5dccb8d85fddb129d6938b994e/setup.native.ts

I am not able to see in documentation how to make this working in version 6.

The updated repository is here https://github.com/MartinCerny-awin/ReactHookForm-ReactNativeTestingLibrary and you can see that in upgrade from version 5 to 6 in the latest commit, it stopped working. 

Data is still submitted correctly in v6 without global.window change. Only errors is not being filled in v6. In v5 form data and errors were not getting filled without global.window

**Comment by bluebill1049** [comment]
@keiya01 would you be able to help on this one?

**Reply by keiya01** [reply]
No, problem👍

**Comment by keiya01** [comment, marked as answer]
I think you need to set `roots` option to `jest.config.js`. https://jestjs.io/docs/en/configuration#roots-arraystring

**Reply by keiya01** [reply]
This problem is caused by `node_modules` is included.
I think if you set this config, you can use a previous config.

**Reply by keiya01** [reply]
Or you can use `transformIgnorePatterns`. https://jestjs.io/docs/en/configuration#transformignorepatterns-arraystring
You can refer our `jest.config.js`. https://github.com/react-hook-form/react-hook-form/blob/master/jest.config.js#L61

**Reply by MartinCerny-awin** [reply]
I have moved the code to src and set roots, but that doesn't make any difference https://github.com/MartinCerny-awin/ReactHookForm-ReactNativeTestingLibrary/commit/d401eef7888d90ef50796dc3112b7066406b7721

**Reply by MartinCerny-awin** [reply]
I have also added `transformIgnorePatterns` https://github.com/MartinCerny-awin/ReactHookForm-ReactNativeTestingLibrary/commit/eee477cfd9d6edbe1202d1aecda6575409704ad5 and that didn't make any difference as well

**Reply by keiya01** [reply]
hmmm, I'm investigating this problem.

**Reply by MartinCerny-awin** [reply]
The same problem is for watch() all the values in it are empty. https://github.com/MartinCerny-awin/ReactHookForm-ReactNativeTestingLibrary/commit/937682e41923b5147def3261fb0bedd005cae424

**Reply by keiya01** [reply]
This problem is caused by `window` object is undefined. To resolve this problem, I added https://github.com/react-hook-form/react-hook-form/blob/master/setup.native.ts. 
But this file is affecting you ...

**Reply by MartinCerny-awin** [reply]
I setting window for the tests, but it doesn't do any difference if i uncomment this https://github.com/MartinCerny-awin/ReactHookForm-ReactNativeTestingLibrary/blob/master/src/__tests__/App-test.js#L7-L12

**Reply by keiya01** [reply]
I found the way to resolve this problem.

First: you need to create `setup.js` like below

https://github.com/react-hook-form/react-hook-form/blob/master/setup.native.ts

Second: you need to import `setup.js` in jest config like below.

https://github.com/react-hook-form/react-hook-form/blob/master/jest.config.js#L65

**Reply by MartinCerny-awin** [reply]
Amazing, it is working! It solves a lot of inconsistencies that I had between tests and real behaviour. https://github.com/MartinCerny-awin/ReactHookForm-ReactNativeTestingLibrary/commit/a2175f46eeee4178054fe37dd307b5da9cc69d64

**Reply by bluebill1049** [reply]
thank you @keiya01! our testing wizard 🧙‍♂️

**Reply by MartinCerny-awin** [reply]
It might be good to add it to documentation. Let me know if I should create one for react-native in https://react-hook-form.com/advanced-usage#TestingForm

**Reply by keiya01** [reply]
Thank you very much @MartinCerny-awin. But I already added this.🥳

**Reply by bluebill1049** [reply]
Thanks @marsinearth anyhow 🙏 feel free to submit PR if you find things need to be improved.
